navigationrail/navigationrail/src/main/java/com/microsoft/device/dualscreen/navigationrail/NavigationRailView.kt

```diff
@@ -0,0 +1,595 @@
+import androidx.activity.ComponentActivity
+import androidx.core.view.animation.PathInterpolatorCompat
+import androidx.customview.view.AbsSavedState
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.lifecycleScope
+import androidx.lifecycle.repeatOnLifecycle
+import androidx.transition.ChangeBounds
+import androidx.transition.Transition
+import androidx.transition.TransitionManager
+import androidx.window.layout.WindowInfoTracker
+import androidx.window.layout.WindowLayoutInfo
+import androidx.window.layout.WindowMetricsCalculator
+import com.google.android.material.navigationrail.NavigationRailMenuView
+import com.google.android.material.navigationrail.NavigationRailView
+import com.microsoft.device.dualscreen.utils.wm.DisplayPosition
+import com.microsoft.device.dualscreen.utils.wm.OnVerticalSwipeListener
+import com.microsoft.device.dualscreen.utils.wm.ScreenMode
+import com.microsoft.device.dualscreen.utils.wm.extractFoldingFeatureRect
+import com.microsoft.device.dualscreen.utils.wm.getWindowVisibleDisplayFrame
+import com.microsoft.device.dualscreen.utils.wm.isFoldingFeatureHorizontal
+import com.microsoft.device.dualscreen.utils.wm.isInDualMode
+import com.microsoft.device.dualscreen.utils.wm.isSeparating
+import com.microsoft.device.dualscreen.utils.wm.locationOnScreen
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+private const val MARGIN_LOWERING_FACTOR = 0.95
+
+/**
+ * A sub class of the [NavigationRailView] that can position its children in different ways when the application is spanned on both screens.
+ * Using the [arrangeButtons] and [setMenuGravity] the children can be split in any way between the two screens.
+ * Animations can be used when changing the arrangement of the buttons on the two screen.
+ */
+class NavigationRailView : NavigationRailView {
+    constructor(context: Context) : super(context) {
+        this.registerWindowInfoFlow()
+    }
+
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
+        this.registerWindowInfoFlow()
+        extractAttributes(context, attrs)
+    }
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
+        context,
+        attrs,
+        defStyleAttr
+    ) {
+        this.registerWindowInfoFlow()
+        extractAttributes(context, attrs)
+    }
+
+    private var screenMode = ScreenMode.DUAL_SCREEN
+    private var job: Job? = null
+    private var windowLayoutInfo: WindowLayoutInfo? = null
+
+    private var topBtnCount: Int = -1
+    private var bottomBtnCount: Int = -1
+
+    private fun normalizeFoldingFeatureRectForView(): Rect {
+        return windowLayoutInfo.extractFoldingFeatureRect().apply {
+            offset(-locationOnScreen.x, 0)
+        }
+    }
+
+    private fun registerWindowInfoFlow() {
+        val activity = (context as? ComponentActivity)
```

Comment:
Can we use `Activity` instead of `ComponentActivity`? Using `ComponentActivity` we would be leaving out developers that use `Activity`. If we can use `Activity` (if there is not a requiement that force us to use `ComponentActivity`), we could allow any child from Activity to be used with the component.

Reply:
Activity doesn't implement LifecycleOwner thus it doesn't has thelifecycleScope.
Also, getLifecycle() is part of the ComponentActivity.